Disk Saving Output Plugin

mohamed1refaie opened the pull request:
Creating an Initial Disk Saving Output Plug-in to Prism engine, addresses #3

mohamed1refaie committed:
Disk Saving Output Plugin

mohamed1refaie committed:
Linting

mohamed1refaie committed:
Refactoring

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+
+//Init func Initialize the disk output plugin
+func (d *Disk) Init(config types.Config, logger zap.Logger) error {
+	d.FileName = config["filename"].(string)
```
Must check if casting is successful, otherwise return error, casting returns a second bool that identifies if casting was correct.

```
d.FileName, stringType = config["filename"].(string)
if !stringType{
       return <error>
}
...
```

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+	d.FileName = config["filename"].(string)
+	d.FilePath = config["filepath"].(string)
+	path.Clean(d.FilePath)
+	d.FilePermission = config["permission"].(os.FileMode)
```
os.FileMode ? 
two things:
1- config is either string or number, so we can't really have os.FileMode.
2- File Permission is in-fact `uint32`,  config will probably cast everything to string only for now and convert to int in code -for now-

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+
+//WriteOnDisk func takes the disk and the transaction
+//that to be written on the disk
+func WriteOnDisk(d *Disk, transaction types.Transaction) {
```
Shouldn't this be a private method receiver for type Disk ?

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+	var err error
+	ack := true
+	d.logger.Info("RECEIVED OUTPUT TRANSACTION...")
+	bytes, _ := ioutil.ReadAll(transaction)
```
don't neglect err :'D

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+	if _, err = os.Stat(d.FilePath); os.IsNotExist(err) {
+		err = os.MkdirAll(d.FilePath, os.ModePerm)
+	}
+	if err == nil {
```
Bos heya detail baseta, bs mmkn t-flip el if condition b7es en el happy path yb2a bara el if ?
I mean, kol checking el errors yb2a `if err != nil` , w myb2ash fe else

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+
+//Disk struct
+type Disk struct {
+	FileName       string
```
Hmm, it may-not be clear but file-name and path should be combined in a single path variable `var/files/images/image_1241.jpg`, etc

This will allow for more flexibility when using dynamic strings.

check https://golang.org/pkg/path/filepath/#Dir to easily check if Dir only exist (not the file)

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+	d.FilePath = config["filepath"].(string)
+	path.Clean(d.FilePath)
+	d.FilePermission = config["permission"].(os.FileMode)
+	d.Transactions = make(chan types.Transaction, 1)
```
Using buffered channel is not a good idea with `Transactions`, because you may send a transaction to the channel (that will get accepted and added to buffer), then you receive a stop signal, so now you'll stop working while there is a transaction that is still in the buffer! this will be lost. and whoever sent this transaction will wait for it forever. causing a deadlock at somepoint.

dummy had them because it was copy pasted but dummy was never serious you knw :"DD

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,103 @@
+				d.logger.Info("Closing Disk Output...")
+				return
+			case transaction, more := <-d.Transactions:
+				if more {
```
No need to check for more, since you close channels after `wg.wait()` which by-itself mean the whole go-routine of receiving transaction has returned.

sherifabdlnaby reviewed (changes requested):
Since this is a non-dummy plugin, we need only to use logging when necessary, all logging here is unnecessary as starting or closing failure/success will be handled by Engine, and other stuff will be spammy.

mohamed1refaie committed:
Requested Changes

1. All logging removed 
2. Error returned if the type casting failed
3. Changed the permission type from os.FIleMode to string
4. Changed WriteOnDisk to be a private method receiver for type disk
5. Every error is handled 
6. FileName and FilePath are now combined in FIlePath
7. Changed Transactions to be an un-buffered channel 
8. Removed the redundant check on "more"
notice: I couldn't flip the if conditions because in this commit there are many checks

mohamed1refaie committed:
go fmt

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -0,0 +1,105 @@
+		}
+	}
+	if err == nil {
+		perm32, errParse := strconv.ParseUint(d.FilePermission, 0, 32)
```
This is static, no need to parse it, and cast it at every write, final value better be calculated at `Init()` instead.

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -29,33 +31,40 @@ func (d *Disk) TransactionChan() chan<- types.Transaction {
+//WriteOnDisk func takes the transaction
 //that to be written on the disk
-func WriteOnDisk(d *Disk, transaction types.Transaction) {
+func (d *Disk) WriteOnDisk(transaction types.Transaction) {
```
still public tho

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -29,33 +31,40 @@ func (d *Disk) TransactionChan() chan<- types.Transaction {
-		d.logger.Info("OUTPUT SUCCESSFUL, Sending Response. ")
-	} else {
-		d.logger.Info("Error in output: ", zap.Error(err))
+		perm32,errParse:=strconv.ParseUint(d.FilePermission,0,32)
```
Since these are static, no need to parse and cast at every write. 
you can cache them by doing this on `Init()`

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -29,33 +31,40 @@ func (d *Disk) TransactionChan() chan<- types.Transaction {
-	bytes, _ := ioutil.ReadAll(transaction)
-	if _, err = os.Stat(d.FilePath); os.IsNotExist(err) {
-		err = os.MkdirAll(d.FilePath, os.ModePerm)
+	bytes, err := ioutil.ReadAll(transaction)
```
You can read bytes only when you're sure you'll write, consider moving the reading part only when you're about to write.

sherifabdlnaby commented on internal/output/disk/disk.go:
```diff
@@ -1,25 +1,27 @@
 package disk
 
 import (
+	"github.com/pkg/errors"
```
why external `github.com/pkg/errors` not stdlib `errors` ? 
weird it is working since it wasn't added to go.mod

sherifabdlnaby reviewed (changes requested):
Small changes sorry man :'D
Just to set the quality we need to target isa :D

mohamed1refaie committed:
Requested Changes #2

sherifabdlnaby commented:
Merged 👍
